### PR TITLE
fixit: add sample for cloudcdn sign url prefix

### PR DESCRIPTION
**Pull request by zhumin8**
add sample for cloudcdn_sign_url_prefix as part of fixit.
Doc page: https://cloud.google.com/cdn/docs/using-signed-urls#programmatically_create_signed_urls_with_a_url_prefix

### CODE REVIEW

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java [hunk 1]
```diff
@@ -75,6 +76,61 @@ public static String getSignature(byte[] privateKey, String input)
   }
   // [END cloudcdn_sign_url]
 
+  // [START cloudcdn_sign_url_prefix]
```
Comment (Sita04) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java:79]: One sample per file. Kindly move this sample to a new file.

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java [hunk 2]
```diff
@@ -75,6 +76,61 @@ public static String getSignature(byte[] privateKey, String input)
...
+  /**
+   * Creates a signed URL with a URL prefix for a Cloud CDN endpoint with the given key. Prefixes
+   * allow access to any URL with the same prefix, and can be useful for granting access broader
+   * content without signing multiple URLs.
+   *
+   * @param requestUrl URL of request
```
Comment (Sita04) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java:85]: Applies to all parameters below.
Consider moving these comments to the main method in the new file and refrain from javadoc style comments.

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java [hunk 3]
```diff
@@ -75,6 +76,61 @@ public static String getSignature(byte[] privateKey, String input)
...
+      String urlPrefix,
+      byte[] key,
+      String keyName,
+      Date expirationTime)
+      throws InvalidKeyException, NoSuchAlgorithmException, IllegalArgumentException {
+
+    if (urlPrefix.contains("?") || urlPrefix.contains("#")) {
+      throw new IllegalArgumentException("urlPrefix must not include query params: " + urlPrefix);
+    }
+    if (!urlPrefix.startsWith("http://") && !urlPrefix.startsWith("https://")) {
```
Comment (Sita04) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java:106]:
```suggestion
    if (!urlPrefix.startsWith("http://")) {
```

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java [hunk 4]
```diff
@@ -75,6 +76,61 @@ public static String getSignature(byte[] privateKey, String input)
...
+   * @return a properly formatted signed URL with URL prefix
+   * @throws InvalidKeyException when there is an error generating the signature for the input key
+   * @throws NoSuchAlgorithmException when HmacSHA1 algorithm is not available in the environment
+   * @throws IllegalArgumentException when urlPrefix string is malformed.
+   */
+  static String signUrlWithPrefix(String requestUrl,
+      String urlPrefix,
+      byte[] key,
+      String keyName,
+      Date expirationTime)
```
Comment (Sita04) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java:100]: Kindly move the parameters to a single line and adding line breaks only if they reach the character limit.

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java (hunk 2 above)
Comment (Sita04) [reply cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java:85]: Example: https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/dlp/snippets/src/main/java/dlp/snippets/DeIdenitfyWithDeterministicEncryption.java#L42-L58

**Review by Sita04 — CHANGES_REQUESTED** [review]

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java [hunk 5]
```diff
@@ -75,6 +76,62 @@ public static String getSignature(byte[] privateKey, String input)
...
+      throws InvalidKeyException, NoSuchAlgorithmException {
+
+    final String algorithm = "HmacSHA1";
+    final int offset = 0;
+    Key key = new SecretKeySpec(privateKey, offset, privateKey.length, algorithm);
+    Mac mac = Mac.getInstance(algorithm);
+    mac.init(key);
+    return Base64.getUrlEncoder()
+        .encodeToString(mac.doFinal(input.getBytes(StandardCharsets.UTF_8)));
+  }
```
[the remark this answers is not in the document]
Comment (Sita04) [reply cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java:132]: You can add multiple region tags for a method. But the guideline is to have one sample per file. So, you would need to copy over the method to the new file.

**Comment by Sita04** [comment]
Thanks for the contribution! 

Kindly also address the Inclusive language warnings.

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhumin8) [reply cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: Got it. That's done in the recent update.

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java (hunk 3 above)
Comment (Sita04) [reply cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: Since we only need to check if the urlPrefix does not start with "http://", there is no need to explicitly check for "https://" as well.
The second condition will be captured as part of the first condition.

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrlWithPrefix.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    String keyPath = "/path/to/key";
+    // The date that the signed URL expires
+    Date expirationTime = getTomorrow();
+    // URL of request
+    String requestUrl = "https://media.example.com/videos/id/main.m3u8?userID=abc123&starting_profile=1";
+    // URL prefix to sign as a string. urlPrefix must start with either http:// or
+    // https:// and should not include query parameters
+    String urlPrefix = "https://media.example.com/videos/";
+
+    //read the key as a base 64 url-safe encoded string, then convert to byte array
```
Comment (Sita04) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrlWithPrefix.java:51]:
```suggestion
    // Read the key as a base64 url-safe encoded string, then convert to byte array
```

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrlWithPrefix.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+    final String algorithm = "HmacSHA1";
+    final int offset = 0;
+    Key key = new SecretKeySpec(privateKey, offset, privateKey.length, algorithm);
+    Mac mac = Mac.getInstance(algorithm);
+    mac.init(key);
+    return Base64.getUrlEncoder()
+        .encodeToString(mac.doFinal(input.getBytes(StandardCharsets.UTF_8)));
+  }
+  // [END cloudcdn_sign_url_prefix]
```
Comment (Sita04) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrlWithPrefix.java:106]: Move to line 108

File: cdn/signed-urls/src/test/java/com/google/cdn/SignedUrlsTest.java [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 Google Inc.
+ * Copyright 2018 Google LLC
```
Comment (Sita04) [inline cdn/signed-urls/src/test/java/com/google/cdn/SignedUrlsTest.java:2]: Kindly revert this change. Headers for existing code shouldn't be modified.

File: cdn/signed-urls/src/test/java/com/google/cdn/SignedUrlWithPrefixTest.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+import static com.google.cdn.SignedUrlWithPrefix.signUrlWithPrefix;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThrows;
+
+import java.util.Base64;
+import java.util.Date;
+import org.junit.Test;
+
+public class SignedUrlWithPrefixTest {
```
Comment (Sita04) [inline cdn/signed-urls/src/test/java/com/google/cdn/SignedUrlWithPrefixTest.java:27]:
```suggestion
@RunWith(JUnit4.class)
public class SignedUrlWithPrefixTest {
```

**Review by Sita04 — CHANGES_REQUESTED** [review]
Thanks for the changes. Few minor nits.

File: cdn/signed-urls/src/test/java/com/google/cdn/SignedUrlWithPrefixTest.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    assertThrows(IllegalArgumentException.class,
+        () -> {
+          signUrlWithPrefix(REQUEST_URL, INVALID_URL_PREFIX_1, KEY_BYTES, KEY_NAME, EXPIRATION);
+        });
+    assertThrows(IllegalArgumentException.class,
+        () -> {
+          signUrlWithPrefix(REQUEST_URL, INVALID_URL_PREFIX_2, KEY_BYTES, KEY_NAME, EXPIRATION);
+        });
+  }
+}
```
Comment (Sita04) [inline cdn/signed-urls/src/test/java/com/google/cdn/SignedUrlWithPrefixTest.java:60]: Kindly add a new line at the end of the file.

**Review by Sita04 — APPROVED** [review]
LGTM. Thanks.
Kindly fix the 2 open comments before merging.

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java (hunk 3 above)
Comment (zhumin8) [reply cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: > The second condition will be captured as part of the first condition.

I disagree, both condition are needed. With only the first condition,  `https://` will be falsely identified as invalid. Simply try`assertFalse("https://".startsWith("http://"));`
Unless you suggest to capture invalid cases with _"not starting with `http`"_, which still is not the same as saying "only starting with `http://` and `https://` are valid"

Comment (Sita04) [reply cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: Apologies for the confusion. The suggestion was to make it`(!urlPrefix.startsWith("http")` which would capture both the conditions.
Unless you have to explicitly put both conditions in there, might be okay to change it.

Comment (Sita04) [reply cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: Actually on a second thought, I think it's better to demonstrate 2 conditions instead of one. That way the customer knows the URL can start with either "http" or "https".
By simplifying the condition, we'll be only checking if it starts with "http".

Hence, resolving this comment. Thank you.

**Comment by msampathkumar** [comment]
Nice work, @zhumin8 @Sita04. LGTM but let me include a TW, to look for any improvements that can be made to the code comments.

@upeeth - PTAL at this request.

**Comment by upeeth** [comment]
Please rephrase as follows the comments used in the sample published at: https://cloud.google.com/cdn/docs/using-signed-urls#go_1 .

// - urlPrefix must start with "https://" and must not include query parameters.
// - key must be in raw form (not base64-encoded) and up to 16 bytes.
// - keyName must match a key added to the backend service or bucket.

**Comment by zhumin8** [comment]
@upeeth 
Regarding the comment style and wording:
- Updated to include key requirement in comment https://github.com/GoogleCloudPlatform/java-docs-samples/pull/8103/commits/2cf9d00222f4802b31dbc73e6eb230e11a33047a
- I don't think Go sample quoted in your comment is a good candidate to follow on wording. The `urlPrefix must start with "https://" ` conflicts with documentation page highlighted [here](https://cloud.google.com/cdn/docs/using-signed-urls#url-prefix:~:text=A%20URLPrefix%20encodes%20a%20scheme%20(either%20http%3A//%20or%20https%3A//)%2C) 
- The current comment style is updated to follow guide and example from @Sita04 (see comment [here](https://github.com/GoogleCloudPlatform/java-docs-samples/pull/8103#discussion_r1196825659)) and in the spirit of avoiding javadoc comments as stated in [SAMPLE_FORMAT.md](https://github.com/GoogleCloudPlatform/java-docs-samples/pull/8103#discussion_r1196825659). So the comments that corresponds to the Go sample given in your comment, is within main method [here](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/2cf9d00222f4802b31dbc73e6eb230e11a33047a/cdn/signed-urls/src/main/java/com/google/cdn/SignedUrlWithPrefix.java#L38-L49). Let me know if this looks okay or the style is conflicting with any wider guidelines we are trying to follow?

**Comment by upeeth** [comment]
Please compare my suggestions with the comments. The changes suggested are purely editorial and not technical. Use "must" instead of "should" etc.

**Comment by Shabirmean** [comment]
> Please compare my suggestions with the comments. The changes suggested are purely editorial and not technical. Use "must" instead of "should" etc.

@zhumin8 - I think we can go ahead with the PR if you can fix the `should` -> `must` suggestion!. Let me know once it's done.

**Comment by zhumin8** [comment]
@upeeth Ah, thanks for the clarification. Updated in https://github.com/GoogleCloudPlatform/java-docs-samples/pull/8103/commits/b82a1f74cd3481a1cefb20e82234cb9f800d1e51.
@Shabirmean Done. PTAL

**Review by Shabirmean — APPROVED** [review]
